File: backend/app/models/blog.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field
from .user import PyObjectId, UserInDB
from bson import ObjectId

class BlogPostBase(BaseModel):
    title: str
    slug: str
    content: str
    excerpt: str
    cover_image: Optional[str] = None
    author_id: str
    category: str
    tags: List[str] = []
    
class BlogPostCreate(BlogPostBase):
    pass

class BlogPostUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    excerpt: Optional[str] = None
    cover_image: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None

class BlogPostInDB(BlogPostBase):
    id: PyObjectId = Field(default_factory=lambda: str(ObjectId()), alias="_id")
    published_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_schema_extra": {
            "example": {
                "_id": "123456789012345678901234",
                "title": "Introduction à React",
                "slug": "introduction-a-react",
                "content": "Contenu de l'article...",
                "excerpt": "Un aperçu de l'article...",
                "cover_image": "https://example.com/image.jpg",
                "author_id": "123456789012345678901234",
                "category": "Frontend",
                "tags": ["React", "JavaScript", "Web"],
                "published_at": "2023-01-01T00:00:00",
                "updated_at": "2023-01-01T00:00:00"
            }
        }
    }

class BlogPostWithAuthor(BlogPostInDB):
    author: dict
    
    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True
    }

class BlogComment(BaseModel):
    id: PyObjectId = Field(default_factory=lambda: str(ObjectId()), alias="_id")
    content: str
    author_id: str
    post_id: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True
    }

File: frontend/src/types/blog.ts
export interface Category {
  _id: string;
  name: string;
  slug: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface Tag {
  _id: string;
  name: string;
  slug: string;
  created_at: string;
  updated_at: string;
}

export interface BlogPost {
  _id: string;
  title: string;
  slug: string;
  content: string;
  excerpt: string;
  cover_image?: string;
  author_id: string;
  category: string | Category; // Peut être un ID de catégorie ou un objet Category complet
  tags: string[] | Tag[]; // Peut être des IDs de tags ou des objets Tag complets
  published_at: string;
  updated_at: string;
  author?: {
    id: string;
    full_name: string;
    email: string;
    role: string;
  };
}

export interface BlogPostCreate {
  title: string;
  slug: string;
  content: string;
  excerpt: string;
  cover_image?: string;
  category: string;
  tags: string[];
  author_id?: string;
}

export interface BlogPostUpdate {
  title?: string;
  content?: string;
  excerpt?: string;
  cover_image?: string;
  category?: string;
  tags?: string[];
}
